test/integration/data/elastic_torch_main.py

```diff
@@ -0,0 +1,136 @@
+
+args = parser.parse_args()
+
+hvd.init()
+
+batch_size = 32
+data = torch.randn(batch_size, 2)
+target = torch.LongTensor(batch_size).random_() % 2
+
+lr = 0.001
+model = torch.nn.Sequential(torch.nn.Linear(2, 2))
+optimizer = torch.optim.SGD(model.parameters(), lr=lr * hvd.size())
+optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
+
+hostname = os.environ.get('HOROVOD_HOSTNAME')
+start_rank = int(os.environ.get('HOROVOD_RANK', 0))
+
+discovery_schedule = json.loads(args.discovery_schedule)
+epoch_to_hosts = {epoch: hosts for epoch, hosts in discovery_schedule if epoch is not None}
+default_hosts = discovery_schedule[-1][1] if len(discovery_schedule) > 0 else []
+
+exit_schedule = json.loads(args.exit_schedule) if args.exit_schedule else {}
+
+
+def check_exit(epoch, batch):
+    key = str((epoch, batch))
+    if key in exit_schedule:
+        ranks_to_exit = exit_schedule[key]
+        if start_rank in ranks_to_exit:
+            raise RuntimeError('check_rank and exit epoch={} batch={} start_rank={} rank={}'
+                               .format(epoch, batch, start_rank, hvd.rank()))
+
+
+def log_state(state):
+    state_dict = {
+        'epoch': state.epoch,
+        'batch': state.batch,
+        'commits': state.commits,
+        'hostname': hostname,
+        'start_rank': start_rank,
+        'rank': hvd.rank(),
+        'size': hvd.size(),
+        'rendezvous': state.rendezvous}
+    with open(args.logfile, 'a') as f:
+        f.write(json.dumps(state_dict) + os.linesep)
+
+
+@hvd.elastic.run
+def train(state):
+    state.rendezvous += 1
+    while state.epoch < args.epochs:
+        print('epoch {} batch {}'.format(state.epoch, state.batch))
+
+        while state.batch < args.batches_per_epoch:
+            check_exit(state.epoch, state.batch)
+
+            optimizer.zero_grad()
+            output = model(data)
+            loss = F.cross_entropy(output, target)
+            loss.backward()
+            optimizer.step()
+
+            state.batch += 1
+            if state.batch % args.batches_per_commit == 0:
+                state.commits += 1
+                state.commit()
+
+        if hvd.rank() == 0:
+            log_state(state)
+
+            current_hosts = epoch_to_hosts.get(state.epoch, default_hosts)
+            next_hosts = epoch_to_hosts.get(state.epoch + 1, default_hosts)
+            if current_hosts != next_hosts:
+                print('host changes: {} -> {}'.format(current_hosts, next_hosts))
+                start = int(time.time())
+                while state._host_messages.empty():
+                    if int(time.time()) - start > 3:
+                        raise TimeoutError('Timed out waiting for notifications from driver.')
+                    time.sleep(0.1)
+
+        state.epoch += 1
+        state.batch = 0
+        state.commits += 1
+        state.commit()
+
+
+def on_state_reset():
+    for param_group in optimizer.param_groups:
```

Comment:
here as well: `state.optimizer`?

Reply:
Same as above.  In this case we would need to reorder some things, since we create state after this callback function.